### PR TITLE
Add README

### PR DESCRIPTION
**Pull request by helixbass**
In this PR:
- do some stuff to try and get ready to publish an initial version, eg adding a README

Based on `more-languages-fixes`

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,25 +1,40 @@
...
 
 [dependencies]
 bstr = "1.1.0"
 bytecount = "0.6"
-clap = { version = "4.3.0", features = ["derive"] }
+clap = { version = "4.3.0", features = ["derive", "wrap_help"] }
 encoding_rs = "0.8.14"
 encoding_rs_io = "0.1.6"
-ignore = { git = "https://github.com/helixbass/ripgrep", rev = "012f2f1" }
+ignore = { package = "tree_sitter_grep_ignore", git = "https://github.com/helixbass/ripgrep", rev = "669ebd3", version = "0.4.20-dev.0" }
```
Comment (helixbass) [inline Cargo.toml:30]: From reading a couple different things all of your dependencies have to be published (vs eg `git` dependencies) in order for you to publish

So I updated the package names/descriptions of these forked/local dependencies (to have a `tree_sitter_grep_*` prefix) and published them as `whatever-dev.0` versions

At that point I was able to go ahead and publish a `0.1.0-dev.0` version

File: Cargo.toml [hunk 2]
```diff
@@ -1,8 +1,23 @@
...
+tree-sitter-grep is a grep-like search tool that
+recursively searches the current directory for a
+tree-sitter query pattern. Like ripgrep, it respects
+gitignore rules.
+"""
+homepage = "https://github.com/helixbass/tree-sitter-grep"
+repository = "https://github.com/helixbass/tree-sitter-grep"
+keywords = ["tree-sitter", "grep", "search"]
+categories = ["command-line-utilities", "development-tools"]
+rust-version = "1.70"
```
[the remark this answers is not in the document]
Comment (peterstuart) [reply Cargo.toml:20]: You probably want to `exclude` some files here: https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields. With `cargo package --list` I see a bunch of the test fixture code included in the crate. (It doesn't really matter in practice, but adds to the file size when downloading the crate, and seems polite 🤷🏻 )

File: Cargo.toml [hunk 3]
```diff
@@ -48,9 +62,6 @@ tree-sitter-swift = "0.3.6"
 tree-sitter-toml = "0.20.0"
 tree-sitter-typescript = "0.20.2"
 
-[lib]
```
[the remark this answers is not in the document]
Comment (peterstuart) [reply Cargo.toml:51]: Yep, makes sense to me.

File: LICENSE-MIT [hunk 4]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2023 Julian Rosse
```
[the remark this answers is not in the document]
Comment (peterstuart) [reply LICENSE-MIT:3]: Meh

File: README.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+And for "gigantic" code-bases where it's eg scanning > 300k lines of code and outputting > 7000 matches,
+I'm seeing it run in say 360ms, which still feels "quite fast"
+
+
+
+
+## Editor integrations
+
+TODO, I believe that @peterstuart has written an initial version of an Emacs plugin
+and I started tinkering with writing a neovim plugin
```
Comment (peterstuart) [inline README.md:254]: It basically works, I should see about packaging it up and submitting it (or at least adding good instructions).

File: src/args.rs [hunk 6]
```diff
@@ -20,41 +20,83 @@ use crate::{
     ArgGroup::new("query_or_filter")
         .multiple(true)
         .required(true)
-        .args(&["path_to_query_file", "query_source", "filter"])
+        .args(&["path_to_query_file", "query_text", "filter"])
```
[the remark this answers is not in the document]
Comment (peterstuart) [reply src/args.rs:23]: Yeah, I think that's an improvement. I think just `query` would also be fine.

File: src/args.rs [hunk 7]
```diff
@@ -20,41 +20,83 @@ use crate::{
...
+    /// By default this is the "first" capture encountered
+    /// in the query source text.
     #[arg(short, long = "capture")]
     pub capture_name: Option<String>,
 
+    /// The target language for matching.
+    ///
+    /// By default all files corresponding to supported
+    /// languages will be searched if the provided query
+    /// can be successfully parsed for that language.
```
Comment (peterstuart) [inline src/args.rs:53]: Is rustfmt doing the wrapping here? (It is doing it with a narrower width than I'd expect.)

**Review by peterstuart — APPROVED** [review]

File: src/args.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (helixbass) [reply src/args.rs:23]: > I think just `query` would also be fine

Ya I have also considered that, ok updated to `--query` but kept the ("internal") `Args` field name as `query_text` and kept the `--help` "hint value" as `QUERY_TEXT`

File: src/args.rs (hunk 7 above)
Comment (helixbass) [reply src/args.rs]: Ah I guess I didn't know it would do that for me, updated to "let it do the line-breaking"

**Pull request merged**